## Analysis of used  car listings

The data originates from the German eBay website eBay Kleinanzeigen and is about used cars that are for sale. The dataset was scraped and uploaded to Kaggle by user orgesleka.

The aim of this project is to clean and analyze the used car listing data.

Import the libraries

In [1]:
import pandas  as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset includes 50000 rows and 20 columns. The dtype column is containing strings as well as ints. Some column include less than 50000 values, such as vehicleType and gearbox.

In [6]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [7]:
columns_copy = autos.columns.copy()

In [8]:
autos.columns = ["date_crawled", "name", "seller", "offer_type", "price", "abtest", "vehicle_type", "registration_year", "gearbox", "power_ps", "model", "odometer", "registration_month", "fuel_type", "brand", "unrepaired_damage", "ad_created", "nr_of_pictures", "postal_code", "last_seen"]

In [9]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

To unify the column  names, they alll have been converted to lower case and an underscore was included to make separate wordsmore readible. This way the collumn names can be quicklly interpreteted and and the data in better structured.

In [11]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [12]:
autos.head(20)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15  2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16  2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18  2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19  2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   

    seller offer_type    price   abtest vehicle_type  registration_year  \
0   privat    Angebot   $5,000  control          bus               2004   
1   privat    Angebot   $8,500  control    limousine               1997   
2   privat    Angebot   $8,990     test    limousine               2009   
3   privat    Angebot   $4,350  control   kleinwagen               2007   
4   privat    Angebot   $1,350     test        kombi               2003   
5   privat    Angebot   $7,900     test          bus               2006   
6   privat    Angebot     $300     test    limousine               1995   
7   privat    Angebot   $1,990  control    limousine               1998   
8   privat    Angebot     $250     test          NaN               2000   
9   privat    Angebot     $590  control          bus               1997   
10  privat    Angebot     $999     test          NaN               2017   
11  privat    Angebot     $350  control          NaN               2000   
12  privat    Angebot   $5,299  control   kleinwagen               2010   
13  privat    Angebot   $1,350  control    limousine               1999   
14  privat    Angebot   $3,999     test   kleinwagen               2007   
15  privat    Angebot  $18,900     test        coupe               1982   
16  privat    Angebot     $350     test        kombi               1999   
17  privat    Angebot   $5,500     test        coupe               1990   
18  privat    Angebot     $300  control          bus               1995   
19  privat    Angebot   $4,150  control          suv               2004   

      gearbox  power_ps     model   odometer  registration_month fuel_type  \
0     manuell       158    andere  150,000km                   3       lpg   
1   automatik       286       7er  150,000km                   6    benzin   
2     manuell       102      golf   70,000km                   7    benzin   
3   automatik        71    fortwo   70,000km                   6    benzin   
4     manuell         0     focus  150,000km                   7    benzin   
5   automatik       150   voyager  150,000km                   4    diesel   
6     manuell        90      golf  150,000km                   8    benzin   
7     manuell        90      golf  150,000km                  12    diesel   
8     manuell         0     arosa  150,000km                  10       NaN   
9     manuell   

The columns seller and offer_type contain mostly one value and are candidates to be dropped.

The column number_of_pictures needs to be investigated further as the first 20 rows of the dataset contain only the value 0.

The columns price and odometer contain numeric data stored as text. 



### Exploring the price and odometer columns

In [13]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(float)

In [14]:
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].astype(float)
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

The unique values of the columns price and odometer_km

In [15]:
print("price:", autos["price"].unique().shape)
print("odometer_km:", autos["odometer_km"].unique().shape) 

price: (2357,)
odometer_km: (13,)


Let's investigate min/max/median/mean in the columns price and odometer_km

In [16]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [17]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Let's investigate the most frequent values in the price and odometer_km columns

In [18]:
autos["price"].value_counts().head(50)

price
0.0       1421
500.0      781
1500.0     734
2500.0     643
1000.0     639
1200.0     639
600.0      531
800.0      498
3500.0     498
2000.0     460
999.0      434
750.0      433
900.0      420
650.0      419
850.0      410
700.0      395
4500.0     394
300.0      384
2200.0     382
950.0      379
1100.0     376
1300.0     371
3000.0     365
550.0      356
1800.0     355
5500.0     340
350.0      335
1250.0     335
1600.0     327
1999.0     322
400.0      321
6500.0     300
1400.0     292
250.0      291
2800.0     291
2300.0     290
1350.0     276
1700.0     268
200.0      266
450.0      265
3800.0     264
3200.0     261
2900.0     256
1450.0     252
4000.0     246
2999.0     242
2600.0     241
1900.0     239
5000.0     239
1650.0     238
Name: count, dtype: int64

In [19]:
autos["price"].value_counts().sort_index(ascending=False).head(60)

price
99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
194000.0      1
190000.0      1
180000.0      1
175000.0      1
169999.0      1
169000.0      1
163991.0      1
163500.0      1
155000.0      1
151990.0      1
145000.0      1
139997.0      1
137999.0      1
135000.0      1
130000.0      1
129000.0      1
128000.0      1
120000.0      2
119900.0      1
119500.0      1
116000.0      1
115991.0      1
115000.0      1
114400.0      1
109999.0      1
105000.0      2
104900.0      1
99900.0       2
99000.0       2
98500.0       1
94999.0       1
93911.0       1
93000.0       2
89900.0       1
89000.0       1
88900.0       1
86500.0       1
85000.0       1
84997.0       1
84000.0       1
Name: count, dtype: int64

In the price column we discovered that the majority of the values are in the lower price ranges. The value 0 is the most frequent.

In [20]:
autos["odometer_km"].value_counts().head(20)

odometer_km
150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: count, dtype: int64

In [21]:
autos["odometer_km"].value_counts().sort_index(ascending=False)

odometer_km
150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
5000.0        967
Name: count, dtype: int64

In the column odometer_km most values are in higher range, and the maximum value 150000, is also the most frequent.

As for the price column values above 350000 are getting less frequent, we will remove these, as there are only very few occurances.

In [22]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()


count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

### Exploring the date columns

In [23]:
(autos["date_crawled"]
    .str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index())


date_crawled
2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: proportion, dtype: float64

In [24]:
(autos["ad_created"]
    .str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index())

ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: proportion, Length: 76, dtype: float64

In [25]:
(autos["last_seen"]
    .str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index())

last_seen
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: proportion, dtype: float64

In [26]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The min and max values of the registration_year seem odd as they are the year 1000 and 9999 and it seems unlikely these values are correct.

Let's decide which registration years are the lowest and highest acceptable values. Since the first cars have been around in the first decades of the 1900s and the listings are from 2016, we'll go with 1900 - 2016.

In [27]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True)

registration_year
2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
          ...   
1939    0.000021
1948    0.000021
1938    0.000021
1953    0.000021
1943    0.000021
Name: proportion, Length: 78, dtype: float64

The majority of the cars listed habe been registere in the early 2000s.

### Explore the brand column

In [28]:
brand_groups = autos["brand"].value_counts(normalize=True)
common_brands = brand_groups[brand_groups > 0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object', name='brand')


In the cell above you can the the most frequent brands appearing in this dataset (Volkswagen, BMW, Opel, Mercedes Benz, Audi and Ford). They make up the groups of brands that are larger than 5%. It is noticable that the majority of these groups are brands from Germany.

In [29]:
aggregate_dict_mean_prices = {}

for row in common_brands:
    selected_rows = autos[autos["brand"] == row]
    mean_price = selected_rows["price"].mean()
    aggregate_dict_mean_prices[row] = mean_price

aggregate_dict_mean_prices

{'volkswagen': 5402.410261610221,
 'bmw': 8332.820517811953,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 3749.4695065890287}

The most frequent brands in the data set are ranging in mean prices from 2975 $ (Opel) to 9336 $ (Audi). The most frequent brand Volkswagen shows a mean price of 5402 $. Overall Audi, BMW and Mercedes are more expensive, Ford and Open are less expensive and Volkswagwn is in between.

In [30]:
mean_prices_series = pd.Series(aggregate_dict_mean_prices)
print(mean_prices_series)

df = pd.DataFrame(mean_prices_series, columns = ["mean_price"])
print(df)

volkswagen       5402.410262
bmw              8332.820518
opel             2975.241935
mercedes_benz    8628.450366
audi             9336.687454
ford             3749.469507
dtype: float64
                mean_price
volkswagen     5402.410262
bmw            8332.820518
opel           2975.241935
mercedes_benz  8628.450366
audi           9336.687454
ford           3749.469507


In [31]:
brand_mean_mileage = {}

for row in common_brands:
    selected_rows = autos[autos["brand"] == row]
    mean_odometer = selected_rows["odometer_km"].mean()
    brand_mean_mileage[row] = int(mean_odometer)
    
brand_mean_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [32]:
brand_mean_mileage = pd.Series(brand_mean_mileage)
print(brand_mean_mileage)

volkswagen       128707
bmw              132572
opel             129310
mercedes_benz    130788
audi             129157
ford             124266
dtype: int64


In [33]:
brand_info = pd.DataFrame(brand_mean_mileage, columns = ["mean_mileage"])
print(brand_info)

               mean_mileage
volkswagen           128707
bmw                  132572
opel                 129310
mercedes_benz        130788
audi                 129157
ford                 124266


In [34]:
brand_info["mean_price"] = mean_prices_series
brand_info

mean_mileage   mean_price
volkswagen           128707  5402.410262
bmw                  132572  8332.820518
opel                 129310  2975.241935
mercedes_benz        130788  8628.450366
audi                 129157  9336.687454
ford                 124266  3749.469507

Even though the mean price in ranging largely for the top brands, the mean_mileage for Volkswagen, BMW, Opel, Mercedes Benz, Audi and Ford remains similar, ranging in between 120.000 and 130.000.

There is a slight trend with the more expensive vehicles having a lower mileage, versus the more affordable vehicles having a higher mileage.